utils/mangler/src/main/scala/aqua/mangler/ManglerState.scala

```diff
@@ -0,0 +1,65 @@
+package aqua.mangler
+
+import cats.Monoid
+
+import scala.annotation.tailrec
+
+case class ManglerState(namesNumbers: Map[String, Int] = Map.empty) {
+
+  private def genName(name: String, n: Int) =
+    s"$name-$n"
+
+  // find unique names that have not yet been used
+  def findNewNames(introduce: Set[String]): (ManglerState, Map[String, String]) = {
+    introduce.foldLeft(this, Map.empty[String, String]) {
```

Comment:
nit: I think this class may implement just `findNewName(name: String)` and renaming for a set could be done in `Mangler` trait just by `traverse`. I see no gain in optimization or code clarity from implementing this method for set directly.